### PR TITLE
How to upload files to AWS S3 using Django Rest Framework

### LINKED ISSUE #1733: How to upload files to AWS S3 using Django Rest Framework
Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

### Proposed title of article
How to upload files to AWS S3 using Django Rest Framework

### Introduction paragraph (2-3 paragraphs):
File upload and download are some of the most performed actions on the web. With the rise of cloud-based services and functions, which includes storage service, companies are moving their files from an on-premises solution to cloud storage mainly for its cheaper cost and convenience. This article shows how to upload and download user-generated files like images and document files to Amazon S3 via Django Rest Framework which is one of Django's most popular Restful API framework.

### Key takeaways: 
At the end of this article, the reader should be able to:
- Understand what type of files are stored on AWS S3
- Have a working Restful API endpoint that connects to AWS S3
- Understand the need for verifying user-generated files
- Upload and download user-generated files from AWS S3

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.
- [AWS Boto3 Python documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-examples.html)

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Damephena**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

Fixes #1733

### CODE REVIEW

File: articles/how-to-upload-files-to-aws-s3-using-django-rest-framework/index.md [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+
+### Prerequisites
+
+1. Comfortable working with Django and Django Rest Framework
+
+### What is AWS S3?
+
+AWS S3 is an acronym for Amazon Web Services Simple Storage Service (AWS S3). It is a cloud-based service by Amazon for object storage.
+
+Object storage is a type of storage where items processed as a data object. Contrary to the traditional method of storing files in the file system hierarchy.
```
Comment (paulodhiambo) [inline articles/how-to-upload-files-to-aws-s3-using-django-rest-framework/index.md:27]: Consider changing "...items processed as a data object" to "...items are processed as a data object"

File: articles/how-to-upload-files-to-aws-s3-using-django-rest-framework/index.md [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+
+In traditional file system, the basic unit of storage is a "file". In AWS S3, the basic unit of storage called a "bucket".
```
Comment (paulodhiambo) [inline articles/how-to-upload-files-to-aws-s3-using-django-rest-framework/index.md:29]: Consider changing "In traditional file system" to "In the traditional file system" and "...storage called a "bucket"." to "...storage is called a "bucket"."

File: articles/how-to-upload-files-to-aws-s3-using-django-rest-framework/index.md [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+```bash
+
+python manage.py runserver
+
+```
+
+Visit the running server on a [web browser](http://localhost:8000/accounts/). We find the sample file uploaded in the **media** folder. This sample file was created when we uploaded it via the API on the browser as shown below:
+
+![media_folder_included_vs_code](../how-to-upload-files-to-aws-s3-using-django-rest-framework/media_included_structure.png)
```
Comment (paulodhiambo) [inline articles/how-to-upload-files-to-aws-s3-using-django-rest-framework/index.md:295]: Consider changing all the image paths by appending `/engineering-education/how-to-upload-files-to-aws-s3-using-django-rest-framework/` before every the image name. For example, in this above image change `media_folder_included_vs_code` to `media_folder_included_vs_code`

**Review by paulodhiambo — CHANGES_REQUESTED** [review]
@Damephena an amazing article. Check out the changes requested changes and ping me when done.

**Comment by Damephena** [comment]
Hello @paulodhiambo, thank you for reviewing. I have made the requested changes.

**Comment by paulodhiambo** [comment]
Hello @Damephena I am seeing three index.md files and I do not understand the reason why they are three.

**Comment by Damephena** [comment]
Hi @paulodhiambo, there's only one index.md file. Perhaps you are mistaking the commits as extra index file.

**Comment by paulodhiambo** [comment]
Check on the files changed tab, you will see there are some index.md files grom other PRs

**Comment by Damephena** [comment]
I just did. 
I had re-ran the `git fetch section`. `git pull --rebase section master` commands I when updated the requested changes in order to be in sync with section's branch.
Perhaps, the reason for seeing the included files from other people's article.

**Comment by paulodhiambo** [comment]
@Damephena always make sure you create a new branch every time you're writing a new article. You shouldn't be working on the master branch.

**Comment by Damephena** [comment]
Noted. Thanks @paulodhiambo

**Comment by paulodhiambo** [comment]
@Damephena amazing article. @hectorkambow you can give this article a final review.

**Comment by hectorkambow** [comment]
@Damephena 👀  Im noticing this PR has multiple files attached to it - please fix this - as it may cause merging issues. 

see our [FAQ ](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md)for more info. 

best option would be to create a new PR with only the needed files attached to it. 

Be sure to do a git pull on your forked repo to get the latest changes. All articles should be their own separate branch on your forked repo to avoid these issues moving forward 👍 

cc @paulodhiambo

**Comment by Damephena** [comment]
Hello @hectorkambow, I will close this PR and open a new one with a different branch. Thanks. 
Would use new branches moving forward.

**Pull request closed without merging**